WORKSHOP clean up optional dependency imports

evrose54 opened the pull request:
# Reviewer Checklist

* [x] Required ***existing tests*** pass (ie full_test.sh, others as appropriate)
* [x] NO REQUIRED ***unit tests*** (explain why not required)
* [x] NO REQUIRED ***integration tests*** (explain why not required)
* [x] Required ***documentation*** added for new/modified functionality
* [x] Required ***release notes*** added for new/modified functionality
* [ ] Required ***updates to other repos*** complete

# Related Issues
fixes NRLMMD-GEOIPS/geoips#338
See [this PR](https://github.com/NRLMMD-GEOIPS/geoips/pull/438) for the GeoIPS implementation of this PR. Updating other repos as necessary to reflect this new implementation. Make sure to merge that branch first before merging this one.

# Testing Instructions
Run geoips/tests/integration_tests/full_test.sh to ensure these changes have passed. They passed on my end.

# Summary
*From GEOIPS#338: 2023-07-19, Clean up optional dependencies*

Some GeoIPS readers include optional dependency statements that are required to read
certain file formats. While keeping these imports as optional is ok, we should clean up
the manner in which this is implemented. To do so, we've created a separate
geoips.testing.context_manager.py script which can handle optional imports scattered
throughout the GeoIPS codebase. This is essentially replacing our old manner of optional
dependencies with a new method that keeps things clean.

    - modified: geoips/plugins/modules/procflows/data_fusion.py

evrose54 committed:
updated release notes and repo for optional dependencies

evrose54 commented on docs/source/releases/v1_12_0a0.rst:
```diff
@@ -0,0 +1,33 @@
+Version 1.11.7a0 (2023-10-06)
+*****************************
+
+* Refactor: Clean up optional dependencies throughout code
```
Don't merge this until [this PR](https://github.com/NRLMMD-GEOIPS/geoips/pull/438) has been merged.

jsolbrig replied:
The parent PR has been merged.

evrose54 committed:
changed release notes date

jsolbrig committed:
Merge branch 'main' into 338-clean-up-optional-dependencies

jsolbrig committed:
Fix import

jsolbrig reviewed (changes requested):
@evrose54 I'm running into a problem with this one. When I try to run `create_plugin_registires` I get the following:

```
> create_plugin_registries 
15_183838 create_plugin_registries.py:792  INTERACTIVE: Skipping module 'interp_scipy' from 'geoips', interface_name is 'None'
15_183838 create_plugin_registries.py:792  INTERACTIVE: Skipping module 'boxdefinitions' from 'geoips', interface_name is 'None'
15_183838 create_plugin_registries.py:792  INTERACTIVE: Skipping module 'interp_pyresample' from 'geoips', interface_name is 'None'
15_183839 create_plugin_registries.py:792  INTERACTIVE: Skipping module 'geostationary_geolocation' from 'geoips', interface_name is 'None'
15_183839 create_plugin_registries.py:792  INTERACTIVE: Skipping module 'hrit_reader' from 'geoips', interface_name is 'None'
15_183839 create_plugin_registries.py:792  INTERACTIVE: Skipping module 'remss_reader' from 'geoips', interface_name is 'None'
15_183839 create_plugin_registries.py:792  INTERACTIVE: Skipping module 'tc_file_naming' from 'geoips', interface_name is 'None'
15_183839 create_plugin_registries.py:296  INTERACTIVE: Writing /local/home/jsolbrig/geoips/github/geoips/geoips/registered_plugins.json
15_183839 create_plugin_registries.py:296  INTERACTIVE: Writing /local/home/jsolbrig/geoips/github/geoips_clavrx/geoips_clavrx/registered_plugins.json
Traceback (most recent call last):
  File "/local/home/jsolbrig/geoips/github/geoips/geoips/utils/context_managers.py", line 21, in import_optional_dependencies
    yield None
  File "/local/home/jsolbrig/geoips/github/data_fusion/data_fusion/plugins/modules/procflows/data_fusion.py", line 50, in <module>
    from geoips_db.utils.database_writes import write_stats_to_database
ModuleNotFoundError: No module named 'geoips_db'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/jsolbrig/anaconda3/envs/geoips-github/bin/create_plugin_registries", line 8, in <module>
    sys.exit(main())
  File "/local/home/jsolbrig/geoips/github/geoips/geoips/create_plugin_registries.py", line 881, in main
    create_plugin_registries(plugin_packages, save_type)
  File "/local/home/jsolbrig/geoips/github/geoips/geoips/create_plugin_registries.py", line 388, in create_plugin_registries
    error_message += parse_plugin_paths(plugin_paths, package, pkg_dir, plugins)
  File "/local/home/jsolbrig/geoips/github/geoips/geoips/create_plugin_registries.py", line 471, in parse_plugin_paths
    error_message += add_module_plugin(
  File "/local/home/jsolbrig/geoips/github/geoips/geoips/create_plugin_registries.py", line 748, in add_module_plugin
    spec.loader.exec_module(module)
  File "<frozen importlib._bootstrap_external>", line 883, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "/local/home/jsolbrig/geoips/github/data_fusion/data_fusion/plugins/modules/procflows/data_fusion.py", line 48, in <module>
    with import_optional_dependencies(__file__):
  File "/home/jsolbrig/anaconda3/envs/geoips-github/lib/python3.10/contextlib.py", line 153, in __exit__
    self.gen.throw(typ, value, traceback)
  File "/local/home/jsolbrig/geoips/github/geoips/geoips/utils/context_managers.py", line 28, in import_optional_dependencies
    getattr(LOG, loglevel)(err_str)
AttributeError: 'Logger' object has no attribute '/local/home/jsolbrig/geoips/github/data_fusion/data_fusion/plugins/modules/procflows/data_fusion.py'
```

This looks like it is likely a problem in the `import_optional_dependencies` context manager.

evrose54 committed:
fixed error in data_fusion procflow where we were using legacy argument for import_optional_dependencies

evrose54 committed:
updated this branch with origin branch

evrose54 committed:
updated release notes for this PR

mindyls committed:
Update and rename v1_12_2a0.rst to v1_13_0a0.rst

mindyls committed:
Update index.rst

evrose54 committed:
Merge branch 'main' into 338-clean-up-optional-dependencies

evrose54 commented:
Weird @jsolbrig. I'm not running into this problem at all. Could run ``create_plugin_registries`` on this PR again just to make sure the problem is still occurring on your end?

evrose54 committed:
fixed conflicts

jsolbrig commented:
Weirdly, I no longer see this issue. This appears to be ready. I'm going to force the CI to re-run, then mark this as ready to merge.

jsolbrig committed:
Remove old CHANGELOG.md file

jsolbrig reviewed (changes requested):
This is ready to go, but the release notes need to be updated to the new format when we're able to do so.

evrose54 commented:
Great; Thanks for trying this out again! I wonder what the problem was, but I'm glad to see this is no longer an issue.

jsolbrig committed:
Merge branch 'main' into 338-clean-up-optional-dependencies

biosafetylvl5 committed:
Update to use brass for release notes

jsolbrig committed:
Move to correct version number

evrose54 committed:
updated release notes for this PR

evrose54 committed:
updated index.rst

mindyls committed:
Merge branch 'main' into 338-clean-up-optional-dependencies

mindyls commented on docs/source/releases/HEADER:
```diff
@@ -0,0 +1,13 @@
+.. dropdown:: Distribution Statement
+
+ | # # # Distribution Statement A. Approved for public release. Distribution unlimited.
+ | # # #
+ | # # # Author:
+ | # # # Naval Research Laboratory, Marine Meteorology Division
+ | # # #
+ | # # # This program is free software: you can redistribute it and/or modify it under
+ | # # # the terms of the NRLMMD License included with this program. This program is
+ | # # # distributed WITHOUT ANY WARRANTY; without even the implied warranty of
+ | # # # MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the included license
+ | # # # for more details. If you did not receive the license, for more information see:
+ | # # # https://github.com/U-S-NRL-Marine-Meteorology-Division/
```
Is this for generating the automated release notes?  Should we use the updated header?

jsolbrig replied:
Yes and yes! This should use the new header.

mindyls replied:
You don't want the long one?

mindyls replied:
Done

mindyls committed:
Update HEADER to short version

mindyls committed:
Moving 1.14.0a0 release notes to latest

To work with updated CI

mindyls committed:
Adding latest release note to index.rst

mindyls committed:
Fixing formatting for yaml release note

mindyls commented on docs/source/releases/HEADER:
```diff
@@ -0,0 +1,4 @@
+.. dropdown:: Distribution Statement
+
+ | # # # This source code is protected under the license referenced at
+ | # # # https://github.com/NRLMMD-GEOIPS.
```
Is this still required?  Is this part of brassy?

mindyls replied:
Does it do anything?

evrose54 replied:
I think we can get rid of that. I added that when initially testing brassy as I thought we needed it but after talking to Gwyn it sounds like we don't.

mindyls committed:
Removing latest.rst and removing latest section from index.rst

mindyls committed:
Add latest.rst to .gitignore

mindyls committed:
Removing v1_14_0a0.rst release note, will be auto-generated.

mindyls commented on docs/source/releases/v1_12_0a0.rst:
```diff
@@ -0,0 +1,16 @@
+ | # # # for more details. If you did not receive the license, for more information see:
+ | # # # https://github.com/U-S-NRL-Marine-Meteorology-Division/
+
+Version 1.12.0a0 (2024-02-12)
+*****************************
+
+No updates.
```
Is there a reason 1.12.0a0 and 1.13.0a0 are both added here with "No updates" ?

mindyls replied:
I put 1.13.0a0 back to its original contents. Can we just delete 1.12.0a0?

evrose54 replied:
Yup! I think this branch was made before any updates happened in v1_13_0a0. Not sure about the no updates part. I thought that was from you but maybe it was from me. Can't remember. I'll remove it nonetheless.

mindyls commented on docs/source/releases/v1_13_0a0.rst:
```diff
@@ -12,48 +12,7 @@
  | # # # https://github.com/U-S-NRL-Marine-Meteorology-Division/
 
-Version 1.13.0a0 (2024-06-07)
+Version 1.13.0a0 (2024-07-10)
```
What happened to these updates?

mindyls committed:
Putting these changes back for 1.13.0a0

I think these are valid - because I reference putting linear_norm back to
matplotlib_linear_norm in 1.13.0. Also, I never reference updating the test scripts,
and they have in fact been updated to use geoips run data_fusion

mindyls commented on docs/source/releases/v1_13_0a0.rst:
```diff
@@ -1,16 +1,7 @@
- | # # # MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the included license
- | # # # for more details. If you did not receive the license, for more information see:
- | # # # https://github.com/U-S-NRL-Marine-Meteorology-Division/
+ | # # # This source code is protected under the license referenced at
+ | # # # https://github.com/NRLMMD-GEOIPS.
```
NOTE: For some reason all of the contents of this release note had been removed.  I put the contents back - I don't know why the contents had been removed, but they seem to still be relevant (matplotlib_linear_norm -> linear_norm, and updating test scripts to geoips run data_fusion. Note I cahnged linear_norm BACK to matplotlib_linear_norm in 1.13.0)

evrose54 replied:
Sounds good! Thanks for doing that. I'm not sure how the release notes were removed either. This branch was created a while ago and that might have something to do with it.

mindyls commented on docs/source/releases/latest/666-final-bug-fixes-prior-to-workshop.yaml:
```diff
@@ -0,0 +1,9 @@
+documentation:
+- description: |
+    Add create_plugin_registries and missing test data to README
+  title: 'Add create_plugin_registries and missing test data to README'
+  related-issue:
+    number: 666
+    url: 'https://github.com/NRLMMD-GEOIPS/geoips'
+  files:
+    modified: ['README.md']
```
This was from a previous PR, before we updated the YAML release note formatting.

mindyls commented on docs/source/releases/index.rst:
```diff
@@ -43,6 +34,7 @@ Version 1.12
 
    v1_12_4
    v1_12_3
+   v1_12_0a0
```
Why did we add a release note for a release that appears to have no updates?

evrose54 replied:
I'll remove that.

biosafetylvl5 committed:
Delete docs/source/releases/HEADER

mindyls commented on docs/source/releases/v1_14_0a0.rst:
Note this was removed, and the only remaining update moved to latest/666...yaml.  Now we can auto-create new release note.

evrose54 committed:
removed old release note with no changes

jsolbrig committed:
Remove repo URL from related-issue tag

Remove repo URL from related-issue tag because it is unneeded and will trigger a new CI run (the real goal).

jsolbrig committed:
Edit to test CI

mindyls committed:
Removing 1.14.0a0 release note from index.rst

Now auto generated.

mindyls commented on docs/source/releases/index.rst:
```diff
@@ -1,31 +1,14 @@
 Release Notes
 *************
 
-Version 1.14
-------------
-
-.. toctree::
-   :maxdepth: 1
-
-   v1_14_0a0
-
```
Do not merge until docs/source/releases/*.rst check is failing.

mindyls replied:
![image](https://github.com/user-attachments/assets/4c17a840-3968-467e-a3c4-a35bf69fb89c)

mindyls reviewed (approved):
Once the docs/source/releases/*.rst check is correctly fialing, I'm good to merge this.

mindyls commented:
![image](https://github.com/user-attachments/assets/2a2e2ce1-bf37-442a-8969-7a013bc92697)

Overriding and merging - intentionally updated docs/soure/releases/*.rst files
